Catch parsing errors and report them through toXlsxEither function

bradediger opened the pull request:
An application I'm working on needed a bit better error reporting when XLSX files failed to parse. This change refactors `toXlsx` to depend on a `toXlsxEither` function that explicitly reports failure.

Test suite is green and it is working for some simple use cases in my application, but I haven't exercised it heavily. Let me know your thoughts on this approach. Thanks!

Brad

qrilka commented:
Yes, something like that was in my plans.
And probably it makes sense to check if it make sense to use `MonadError` (like in http://hackage.haskell.org/package/mtl-2.2.1/docs/Control-Monad-Except.html#g:3) but I'm not sure about the compatibility with GHC 7.8

bradediger commented:
Thanks. I'm open to any way to report these errors, but I'm not too familiar with what benefit `MonadError` might provide. 

In an application, I might use a `MonadError ParseError m` constraint to abstract over the concrete type of errors, but in this library there's only one concrete monad in use: `Either ParseError`. There was a bit of effort involved in refactoring towards a value-based error reporting approach, so if any of that could be made more general, I'm certainly open to hearing about it.

qrilka commented on src/Codec/Xlsx/Parser.hs:
```diff
@@ -136,69 +149,60 @@ extractCellValue _ "b" "0" = [CellBool False]
-getComments :: Zip.Archive -> FilePath -> Maybe CommentTable
-getComments ar fp = listToMaybe =<< fromCursor <$> xmlCursor ar fp
+getComments :: Zip.Archive -> FilePath -> Either ParseError (Maybe CommentTable)
+getComments ar fp = (listToMaybe . fromCursor =<<) <$> xmlCursor ar fp  --listToMaybe =<< fromCursor <$$> xmlCursor ar fp
```
commented out but not deleted?

bradediger replied:
Oops!! Fixing.

qrilka commented on src/Codec/Xlsx/Parser.hs:
```diff
@@ -136,69 +149,60 @@ extractCellValue _ "b" "0" = [CellBool False]
-xmlCursor :: Zip.Archive -> FilePath -> Maybe Cursor
-xmlCursor ar fname = parse <$> Zip.findEntryByPath fname ar
+xmlCursor :: Zip.Archive -> FilePath -> Either ParseError (Maybe Cursor)
+xmlCursor ar fname = maybe (Right Nothing) parse $ Zip.findEntryByPath fname ar
```
and for this `hush` could be used - http://hackage.haskell.org/package/errors-2.0.0/docs/Control-Error-Util.html#v:hush

bradediger replied:
If you're ok with pulling in a new `errors` dependency for that, I can make the change. I had used `note` somewhere else as well but didn't want to add the dependency just for that.

qrilka replied:
I haven't used `errors` much but there are some useful utities there which could be handy if `xlsx` will start getting saner error reporting support

qrilka commented on src/Codec/Xlsx/Parser.hs:
```diff
@@ -54,68 +67,68 @@ data WorksheetFile = WorksheetFile { wfName :: Text
-    condFormtattings = M.fromList . map unCfPair  $ cur $/ element (n"conditionalFormatting") >=> fromCursor
+             -> Either ParseError Worksheet
+extractSheet ar sst wf = do
+  let  file = fromJust $ Zip.fromEntry <$> Zip.findEntryByPath (wfPath wf) ar
```
please use at least `fromJustNote` but when there is `MissingFile` it makes sense to use it I guess

bradediger replied:
The `fromJust` was existing and I didn't even notice it... I'll look more closely at this.

qrilka replied:
that code is a couple of years old at least so there are parts to improve :)

qrilka commented:
@bradediger I like your PR but I think `MonadError` will make it more readable with explicit `throwError` and removing not quite so obvious use of bifunctors.
Are you OK with going that way?
And for GHC 7.8 there is `mtl-compat`

qrilka commented:
And using just `type Parser = Either ParseError` would be enough at least for the current needs

bradediger commented:
@qrilka Yes, I'll look into trying to refactor in that direction. And your type synonym was exactly where I started, so I think we have an idea of how to move forward. I'll let you know.

Thanks for your review!

qrilka commented:
Thanks for you effort, it is an important improvement for the library

qrilka commented on src/Codec/Xlsx/Parser.hs:
```diff
@@ -39,23 +41,22 @@ import           Codec.Xlsx.Types.Internal.CustomProperties  as CustomProperties
-toXlsxEither :: L.ByteString -> Either ParseError Xlsx
-toXlsxEither bs = do
-  ar <- first (const InvalidZipArchive) $ Zip.toArchiveOrFail bs
+toXlsxOrError :: MonadError ParseError m => L.ByteString -> m Xlsx
```
@bradediger but why so polymorphic? My last proposal was to stay with `Either` but add a type synonym and use `MonadError` instance for it. Users could use e.g. `hoistEither` to lift `Either`

bradediger replied:
@qrilka Actually I started out with that approach, but ended up chasing the `MonadError` types down into the core. I had thought your suggestion was that it would be more intention-revealing if the core functions used `throwError` etc. rather than working directly with Either. Hence the new `whenXThrow` combinators below instead of `note`, etc.

Are you suggesting that we keep the inner functions in this module specialized to Either? The big advantage I saw of the rewrite I just made is that the user can determine how errors are to be reported, as long as there is a `MonadError` instance.

Sorry, perhaps I'm not clear on exactly what you're suggesting. I haven't worked a ton with `MonadError` so it's quite possible I'm missing an easier way.

qrilka replied:
My idea was just to use `MonadError` instance for `Either`.
As for that option to change how errors get reported - that could be done e.g. with `hoistEither` with no problem.

bradediger replied:
OK, I think I see what you're saying. I'll try that.

qrilka commented on src/Codec/Xlsx/Parser.hs:
```diff
@@ -206,3 +209,14 @@ getRels ar fp = do
+
+-- | Lift an 'Either' value into any 'MonadError'.
+whenLeftThrow :: MonadError e m => Either l r -> (l -> e) -> m r
+whenLeftThrow v e = either (throwError . e) return v
```
if we'll drop `MonadError` it will be just `mapLeft`

qrilka commented on src/Codec/Xlsx/Parser.hs:
```diff
@@ -206,3 +209,14 @@ getRels ar fp = do
+--
+--    Just 3 `whenNothingThrow` SomeError
+whenNothingThrow :: MonadError e m => Maybe a -> e -> m a
+whenNothingThrow v e = maybe (throwError e) return v
```
what about using `note`?

qrilka commented:
I could explain a bit more about my points for yesterday:

1. using `throwError` instead of bifunctor functors would make code more readable IMO
2 it makes sense to add a type synonym like `type Parser = Either ParserError`

bradediger commented:
Hi @qrilka,
I just pushed a much more low-touch approach that I think is in the spirit of what you were asking.

* Added `type Parser = Either ParseError`
* Removed bifunctors (replaced with `throwError`, `note`, `either`, and one use of `Control.Arrow.left` since we already use Arrow)
* Replaced `fromJust` with explicit errors

There are a few more uses of `head` that worry me a little bit... do you want those pushed into explicit errors as well?

Let me know what you think of this direction.

qrilka commented on src/Codec/Xlsx/Parser.hs:
```diff
@@ -54,68 +72,70 @@ data WorksheetFile = WorksheetFile { wfName :: Text
+extractSheet ar sst wf = do
+  let filePath = wfPath wf
+  file <- note (MissingFile filePath) $ Zip.fromEntry <$> Zip.findEntryByPath filePath ar
+  cur <- either (\_ -> throwError $ InvalidFile filePath) (return . fromDocument) $ 
```
what about
```haskell
cur <- fromDocument <$> parseLBS def file `catchError` (\_ -> throwError $ InvalidFile filePath)
```
?
A bit more readable IMO

bradediger replied:
I don't think that would work, because the monads are different (`Either SomeException` vs. `Either ParseError`). Any suggestions? It's possible I'm missing something here.

qrilka replied:
oh, yes, exactly, what about `left` then from `Control.Arrow`? :)
it has a bit more telling name than `first` from bifunctors

qrilka commented on src/Codec/Xlsx/Parser.hs:
```diff
@@ -136,69 +156,65 @@ extractCellValue _ "b" "0" = [CellBool False]
-xmlCursor :: Zip.Archive -> FilePath -> Maybe Cursor
-xmlCursor ar fname = parse <$> Zip.findEntryByPath fname ar
+xmlCursor :: Zip.Archive -> FilePath -> Parser (Maybe Cursor)
+xmlCursor ar fname = maybe (Right Nothing) parse $ Zip.findEntryByPath fname ar
```
what about use `note` in `xmlCursorRequired` and `hush` to transform it into `Maybe` in this function? We have `Either` from `parseLBS` and it doesn't look cool to turn it into `Maybe` and then back into `Either`

bradediger replied:
The problem I was having with this is that `xmlCursorRequired` can fail in two different ways: `MissingFile` or `InvalidFile`. We only want to catch `MissingFile` for the purpose of returning Nothing; if the file is present but invalid, the `InvalidFile` should still occur.

bradediger replied:
That sounds like a great solution. If you are up for doing it I am absolutely :+1:. If you don't have the time I can probably find time over the weekend.

qrilka replied:
OK, so I'm merging it and will add follow up commits a bit later in the evening, thank you

bradediger replied:
Thanks @qrilka !

qrilka commented on src/Codec/Xlsx/Parser.hs:
```diff
@@ -136,69 +156,65 @@ extractCellValue _ "b" "0" = [CellBool False]
-    Just c ->
-      let [sst] = fromCursor c in sst
+getSharedStrings  :: Zip.Archive -> Parser SharedStringTable
+getSharedStrings x = head . fromCursor <$> xmlCursorRequired x "xl/sharedStrings.xml"
```
that looks neat :+1:

qrilka commented:
Thanks @bradediger it looks great, after resolving that minor issue I will merge it.
And with upcoming solution for #47 it will go into 0.2.4 release

bradediger commented:
@qrilka thanks for your feedback, will get these changes made soon!
